src/hip/general_tensor_reorder_sol.cpp

```diff
@@ -0,0 +1,214 @@
+                                               uint32_t order_2_,
+                                               uint32_t order_3_)
+    : data_type(data_type_),
+      dim_0(dim_0_),
+      dim_1(dim_1_),
+      dim_2(dim_2_),
+      dim_3(dim_3_),
+      order_0(order_0_),
+      order_1(order_1_),
+      order_2(order_2_),
+      order_3(order_3_)
+{
+    if(data_type == miopenInt8x4)
+        MIOPEN_THROW("These data type are not supported");
+    num_cu                 = ctx.GetStream().GetMaxComputeUnits();
+    std::size_t data_size  = miopen::GetTypeSize(data_type);
+    kernel_param_heuristic = tensor_reorder::HeuristicGet(data_size, dim_0, dim_1, dim_2, dim_3);
+}
+
+solver::KernelInfo GeneralReorderSolution::GetKernel() const
+{
+    std::size_t block_size = TENSOR_REORDER_BLOCK_SIZE;
+    uint32_t pixel_total   = dim_0 * dim_1 * dim_2 * dim_3;
+    uint32_t dim_total     = (pixel_total + block_size * kernel_param_heuristic.tile_x - 1) /
+                         (block_size * kernel_param_heuristic.tile_x);
+    std::size_t grid_size = dim_total;
+
+    std::string kernel_name = GetKernelName();
+    solver::KernelInfo kernel;
+    kernel.kernel_file = "general_tensor_reorder.cpp";
+    kernel.kernel_name = kernel_name;
+    kernel.g_wk.clear();
+    kernel.g_wk.push_back(grid_size * block_size);
+    kernel.g_wk.push_back(1);
+    kernel.g_wk.push_back(1);
+    kernel.l_wk.clear();
+    kernel.l_wk.push_back(block_size);
+    kernel.l_wk.push_back(1);
+    kernel.l_wk.push_back(1);
+
+    MIOPEN_LOG_I2("GeneralReorderSolution use kernel: " + kernel_name);
+
+    return kernel;
+}
+
+std::vector<OpKernelArg> GeneralReorderSolution::GetKernelArg() const
+{
+    std::size_t block_size = TENSOR_REORDER_BLOCK_SIZE;
+    uint32_t pixel_total   = dim_0 * dim_1 * dim_2 * dim_3;
+    uint32_t dim_total     = (pixel_total + block_size * kernel_param_heuristic.tile_x - 1) /
+                         (block_size * kernel_param_heuristic.tile_x);
+    std::size_t grid_size = dim_total;
+
+    magic_div_u32_t magic_stride0 = magic_div_u32_gen(dim_1 * dim_2 * dim_3);
+    magic_div_u32_t magic_stride1 = magic_div_u32_gen(dim_2 * dim_3);
+    magic_div_u32_t magic_stride2 = magic_div_u32_gen(dim_3);
+
+    std::vector<OpKernelArg> opArgs;
+    opArgs.emplace_back(0); // placeholder
+    opArgs.emplace_back(0); // placeholder
+    opArgs.emplace_back(dim_0);
+    opArgs.emplace_back(dim_1);
+    opArgs.emplace_back(dim_2);
+    opArgs.emplace_back(dim_3);
+    opArgs.emplace_back(static_cast<uint32_t>(grid_size));
+    opArgs.emplace_back(dim_total);
+    opArgs.emplace_back(magic_stride0.magic);
+    opArgs.emplace_back(static_cast<uint32_t>(magic_stride0.shift));
+    opArgs.emplace_back(magic_stride1.magic);
+    opArgs.emplace_back(static_cast<uint32_t>(magic_stride1.shift));
+    opArgs.emplace_back(magic_stride2.magic);
+    opArgs.emplace_back(static_cast<uint32_t>(magic_stride2.shift));
```

Comment:
Any guarantees that these casts to uint32_t are safe? If not, then let's `if(grid_size != static_cast<uint32_t>(grid_size)) MIOPEN_THROW...` etc.

Reply:
Answer required, thanks.

Reply:
member `shift` in `magic `like `magic_stride1.shift` is a `uint8_t` value smaller than 0x1F, I think it's safe to casting from it to `uint32_t`.

Reply:
`magic.shift` is `uint8_t` and smaller than 0x1F, so I think it's safe to cast to uint32_t.

Reply:
🟢 Resolved!